### PR TITLE
Additional checks in BF kNN, NaN/Inf checker and train_test_split

### PR DESCRIPTION
**Pull request by Alexsandruss**
# Description
Additional parameter and integer overflow checks in Brute Force k-Nearest Neighbors, NaN/Inf checker and train_test_split

### CODE REVIEW

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_classification_model_impl.cpp [hunk 1]
```diff
@@ -64,7 +64,8 @@ size_t Model::getNumberOfFeatures() const
 
 services::Status Parameter::check() const
 {
-    DAAL_CHECK_EX(this->nClasses > 0, services::ErrorIncorrectParameter, services::ParameterName, nClassesStr());
+    DAAL_CHECK_EX(this->nClasses > 1, services::ErrorIncorrectParameter, services::ParameterName, nClassesStr());
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_classification_model_impl.cpp:67]: it is necessary to add badarg tests for these parameters. we can't work with one class for this algorithm?

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 2]
```diff
@@ -246,6 +251,7 @@ protected:
             int * indices = indexesBlock.get();
 
             const size_t size = blockSize * k * sizeof(*indices);
+            DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, blockSize * sizeof(*indices), k);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:254]: I think you can do this check before the for.

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 3]
```diff
@@ -256,6 +262,7 @@ protected:
             FPType * distances = distancesBlock.get();
 
             const size_t size = blockSize * k * sizeof(FPType);
+            DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, blockSize * sizeof(FPType), k);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:265]: I think you can do this check before the for.

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 4]
```diff
@@ -256,6 +262,7 @@ protected:
...
             daal::services::internal::daal_memcpy_s(distances, size, kDistances, size);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:266]: Check copy status?

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i [hunk 5]
```diff
@@ -246,6 +251,7 @@ protected:
             int * indices = indexesBlock.get();
 
             const size_t size = blockSize * k * sizeof(*indices);
+            DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, blockSize * sizeof(*indices), k);
             daal::services::internal::daal_memcpy_s(indices, size, kIndexes, size);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:255]: Check copy status?

**Review by PetrovKP — COMMENTED** [review]
Missing all checks for converting to lower type. For example:
[Here](https://github.com/Alexsandruss/oneDAL/blob/dev/bf-knn-errors-handling/cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i#L323) you need check:
```c++
DAAL_ASSERT(idx[j] + j1<= static_cast<size_t>(services::internal::MaxVal<int>::get()));
```
[Here ](https://github.com/Alexsandruss/oneDAL/blob/dev/bf-knn-errors-handling/cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i#L349) you need check:
```c++
DAAL_ASSERT(kk <= static_cast<size_t>(services::internal::MaxVal<int>::get()));
```
etc.

Check in all places.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 6]
```diff
@@ -247,6 +247,8 @@ double computeSumSOAAVX512Impl(NumericTable & table, bool & sumIsFinite)
     daal::threader_for_break(nCols, nCols, [&](size_t i, bool & needBreak) {
         double * localSum     = tlsSum.local();
         bool * localNotFinite = tlsNotFinite.local();
+        DAAL_CHECK_MALLOC_THR(localSum);
```
Comment (PetrovKP) [inline cpp/daal/src/data_management/finiteness_checker.cpp:250]: You do not use or check the status `safeStat` after the end of the for. You need to add status to the function `computeSumSOAAVX512Impl` - or return

File: cpp/daal/src/data_management/train_test_split.cpp [hunk 7]
```diff
@@ -96,6 +96,7 @@ services::Status generateShuffledIndicesImpl(const NumericTablePtr & idxTable, c
     const size_t n         = idxTable->getNumberOfRows();
     const size_t stateSize = rngStateTable->getNumberOfRows();
     // number of generated uints: 1.5x of n for reserve
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, n / 2, 3);
     const size_t nRandomUInts = n * 3 / 2;
```
Comment (PetrovKP) [inline cpp/daal/src/data_management/train_test_split.cpp:100]: I would add a literal, because these operations can happen in int type
```suggestion
    const size_t nRandomUInts = n * 3lu / 2lu;
```

Comment (PetrovKP) [reply cpp/daal/src/data_management/train_test_split.cpp:100]: And you need to change the order - as done in the check.

Comment (Alexsandruss) [reply cpp/daal/src/data_management/train_test_split.cpp]: Fixed

File: cpp/daal/src/data_management/finiteness_checker.cpp (hunk 6 above)
Comment (Alexsandruss) [reply cpp/daal/src/data_management/finiteness_checker.cpp:250]: Fixed

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 5 above)
Comment (Alexsandruss) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i]: Check added

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 4 above)
Comment (Alexsandruss) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i]: Check added

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 3 above)
Comment (Alexsandruss) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:266]: Fixed

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i (hunk 2 above)
Comment (Alexsandruss) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i:255]: Fixed

File: cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_classification_model_impl.cpp (hunk 1 above)
Comment (Alexsandruss) [reply cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_classification_model_impl.cpp]: badarg added, nClasses = 1 is non default and obviously incorrect if set by user.

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/eb035469-1daa-f10f-b869-8cdcd4b723a1

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 8]
```diff
@@ -449,6 +453,7 @@ bool checkFinitenessSOAAVX512Impl(NumericTable & table, bool allowNaN)
 
     daal::threader_for_break(nCols, nCols, [&](size_t i, bool & needBreak) {
         bool * localNotFinite = tlsNotFinite.local();
+        DAAL_CHECK_MALLOC_THR(localNotFinite);
```
Comment (PetrovKP) [inline cpp/daal/src/data_management/finiteness_checker.cpp:456]: After check in threading for, need make detach of `safeStat` status:
See for example [here](https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/kernel_function/kernel_function_linear_dense_default_impl.i#L154).

**Review by PetrovKP — COMMENTED** [review]
Internal falls need to be fixed.

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/data_management/finiteness_checker.cpp (hunk 8 above)
Comment (Alexsandruss) [reply cpp/daal/src/data_management/finiteness_checker.cpp:460]: Fixed. Test fall are fixed too but waiting for CI to approve it

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/eb04b514-b156-f1e4-bde2-8cdcd4b723a1

**Review by PetrovKP — APPROVED** [review]
